Document how to add session middleware back

mltsy opened the pull request:
Adds documentation to the guides about how to get session middleware back when using the api_only flag.  Without this, it's not clear that session middleware has special cases to handle with the `api_only` flag.

Hopefully will help people avoid this problem: http://stackoverflow.com/questions/15342710/adding-cookie-session-store-back-to-rails-api-app/15363386#15363386

Commit:
Document how to add session middleware back

Without this, it's not clear that session middleware has special cases to handle with the `api_only` flag

betesh commented:
:+1:

gmcgibbon commented on guides/source/api_app.md:
````diff
@@ -332,6 +332,28 @@ will be:
+```ruby
+config.session_store :cookie_store, key: '_interslice_session'
+config.middleware.use ActionDispatch::Cookies # Required for all session management
+config.middleware.use ActionDispatch::Session::CookieStore, config.session_options
````
As far as I can tell, we should be referencing `config.action_controller.session_options` and `config.action_controller.session_store` as mentioned [here](https://github.com/rails/rails/blob/d41ec8180d3df391e5fc0189f499d8689ca27e03/guides/source/configuring.md#configuring-middleware) but that doesn't seem to set `session_options` properly. 😕

mltsy replied:
Interesting... yeah I'm not sure why those don't work... I wonder if those are writers only... ? (that doesn't make a lot of sense...), and to be honest I can't remember how I found these values - I assume I just found them in the default middleware stack implementation: https://github.com/rails/rails/blob/master/railties%2Flib%2Frails%2Fapplication%2Fdefault_middleware_stack.rb#L62

It does seem to make sense to change them to match the docs, but if that doesn't work... then I guess it doesn't! 😉

gmcgibbon reviewed (commented):
Still relevant in Rails 6 d41ec8180d3df391e5fc0189f499d8689ca27e03.

mltsy commented on guides/source/api_app.md:
```diff
@@ -332,6 +332,28 @@ will be:
+- `ActionDispatch::Session::MemCacheStore`
+
+The trick to adding these back in is that, by default, they are passed `session_options`
+when added (including the session key), so you can't just add a `session_store.rb` initializer, add
```
I see how this is confusing.  IIRC, the point of this comment is not to do with where you add the middleware (which should be in `application.rb`), but rather that the conventional `session_store.rb` initializer won't work if you are adding these middlewares in manually, because you have to pass session_options to these middlewares, and they won't have been set in the correct scope yet if you use the conventional session_store.rb initializer to set them.  Therefore you have to manually set them in application.rb before "using" the middleware, so that you can pass them to the middleware.

mltsy commented:
I've addressed the comments here, not sure if that was the only think keeping this from being merged - can I reopen this?  Or should I start a new PR?

mltsy commented:
Looks like rails-bot isn't going to help me reopen, so I'll create a new one...
Aw, shoot - almost got exactly 10,000 PRs later:  https://github.com/rails/rails/pull/38098

godfreymutebi commented:
- I have tried the above solutions but non is working. I have to apps running on different domain, the create react app and the rails 6 api only app. what confuses me is that i added activeadmin which works well and it uses sessions which i added back. Am doing an ecommerce app.
- Here is my application.rb
` config.middleware.use Rack::MethodOverride
    config.middleware.use ActionDispatch::Flash
    config.middleware.use ActionDispatch::Cookies
    config.middleware.use ActionDispatch::Session::CookieStore`

- Below is how i want to set sessions from
`module CurrentCart
  private

  def set_cart
    @cart = Cart.find(session[:cart_id])
  rescue ActiveRecord::RecordNotFound
    @cart = Cart.create
    session[:cart_id] = @cart.id
  end
end
`

- I  want to get the value of @cart in session if does not exist it proceeds and creates a new cart and stores the id in sessions.  Any help will be appreciated

cart commented:
> I want to get the value of @cart in session if does not exist it proceeds and creates a new cart and stores the id in sessions. Any help will be appreciated

You can just ask me for the value of `@cart`. I currently still exist.
